Fix physics events being interpreted twice for nodes in canvas layer

Sauermann opened the pull request:
resolve #64652

Currently `GodotPhysicsDirectSpaceState2D::intersect_point` returns, when given the default viewport canvas layer as argument, also `ShapeResult`s that belong to other canvas layers within the current viewport.
Since canvas layer of these nodes can have a different transform, they may not intersect with the given point.

This patch introduces an additional parameter to `PhysicsPointQueryParameters2D`, which governs how `canvas_instance_id == 0` should be handled.

Updated 2022-12-12: Fix merge conflict

Sauermann committed:
Fix physics events being interpreted twice when in canvas layer

Sauermann commented:
I have updated the PR according to the discussion with @rburing.
This version is backwards-compatible for handling multiple canvas layers which have the same transform.

rburing commented:
I'd prefer to just remove the special treatment of `canvas_instance_id == 0` in `intersect_point` (instead of making it optional), since that is pretty much incompatible with the concept that canvas layers can be transformed. IIRC that's how this PR was originally (minus the docs update), and I prefer that (with docs) to the current (more hacky) variant.

That would break compatibility for people using `intersect_point` with areas/bodies in several canvas layers (and expecting to hit all of them in one query). I'd say it's worth it for fixing the bug and keeping the API clean. (We could make a note about the behavior change in the blog post of the beta release.)

Sauermann commented:
The original version of the PR can be seen at https://github.com/godotengine/godot/commit/cbbb2af52fd640c3bf2dba6f9b93714e36223746 (not sure how long github keeps this cached version).

I have reverted the PR to the original version and updated the docs to additionally describe the case `canvas_instance_id == 0`.

akien-mga commented:
Thanks!
